std/range/primitives.d

```diff
@@ -176,6 +170,46 @@ template isInputRange(R)
     static assert( isInputRange!(inout(int)[]));
 }
 
+/**
+Helper function for $(D isInputRange). Can also be used to statically
+assert that a user-defined type is an input range.
```

Comment:
`isInputRange` works just fine for statically asserting that a user-defined type is an input range. `checkInputRange` doesn't help with that at all. So, the docs really don't show how this is different from `isInputRange`. I think that the difference between the two should be made much clearer here. In particular, I think that it needs to be clear than `isInputRange` is used in template constraints (where you don't want to see error messages from inside of the traits you're using), whereas `checkInputRange` is used specifically on your user-defined types to report _why_ they're failing `isInputRange`. And given the risk of confusion between `isInputRange` and `checkInputRange`, I'm inclined to think that we should find a way to do this such that `checkInputRange` _isn't_ a function which returns `bool` and cannot possibly be used in a template constraint. I think that what you have going here is a great idea, but I think that we need to make sure that it's not going to cause increased confusion, and as long as it's possible to use it instead of `isInputRange` in a template constraint, I think that that's a real risk. They need to be distinct. And as it stands, they're not distinct enough in either the documentation or in how they can be used.